etna/datasets/hierarchical_structure.py

```diff
@@ -0,0 +1,179 @@
+from collections import defaultdict
+from itertools import chain
+from queue import Queue
+from typing import DefaultDict
+from typing import Dict
+from typing import List
+from typing import Optional
+from typing import Union
+
+import scipy
+from scipy.sparse import lil_matrix
+
+from etna.core import BaseMixin
+
+
+class HierarchicalStructure(BaseMixin):
+    """Represents hierarchical structure of TSDataset."""
+
+    def __init__(self, level_structure: Dict[str, List[str]], level_names: Optional[List[str]] = None):
+        """Init HierarchicalStructure.
+
+        Parameters
+        ----------
+        level_structure:
+            Adjacency list describing the structure of the hierarchy tree (i.e. {"total":["X", "Y"], "X":["a", "b"], "Y":["c", "d"]}).
+        level_names:
+            Names of levels in the hierarchy in the order from top to bottom (i.e. ["total", "category", "product"]).
+            If None is passed, level names are generated automatically with structure "level_<level_index>".
+        """
+        self._num_nodes = 0
+        self._hierarchy_root: Optional[str] = None
+
+        self.level_structure = level_structure
+
+        self._find_graph_structure(level_structure)
+        hierarchy_levels = self._find_hierarchy_levels(level_structure)
+        tree_depth = len(hierarchy_levels)
```

Comment:
Create static method `_get_level_names(level_names, hierarchy_levels) -> List[str]` and put all the logic from the constructor there